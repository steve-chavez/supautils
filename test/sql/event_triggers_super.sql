-- skipped event triggers are logged
set supautils.log_skipped_evtrigs = true;
\echo

-- a superuser can create an event trigger
set role postgres;
\echo

create function super_show_current_user()
    returns event_trigger
    language plpgsql as
$$
begin
    raise notice 'Executing superuser event trigger: current_user is %', current_user;
end;
$$;
\echo

create event trigger event_trigger_1 on ddl_command_end
execute procedure super_show_current_user();
\echo

-- the event trigger is owned by the superuser
select evtowner::regrole from pg_event_trigger where evtname = 'event_trigger_1';

-- the superuser evtrig will execute for superuser
create table super_thing();
\echo

-- another superuser can create an event trigger
create role super2 superuser;
set role super2;
create function super2_show_current_user()
    returns event_trigger
    language plpgsql as
$$
begin
    raise notice 'Executing super2 event trigger: current_user is %', current_user;
end;
$$;
create event trigger super2_event_trigger on ddl_command_end
execute procedure super2_show_current_user();
\echo

-- only one superuser evtrig will execute for superuser (the one he owns)
create table super2_thing();
\echo

-- super2 evtrigs won't run here, only the postgres owned evtrigs
set role postgres;
create table super_duper_thing();
\echo

-- super2 cleanup
drop event trigger super2_event_trigger;
drop function super2_show_current_user;
drop table super2_thing;
drop role super2;
\echo

-- A regular user is not able to create event trigger using a superuser function
set role privileged_role;
\echo

create event trigger event_trigger_2 on ddl_command_end
execute procedure super_show_current_user();
\echo

-- A reserved role will execute the superuser evtrig
set role supabase_storage_admin;
\echo

create table storage_stuff();
\echo

drop table storage_stuff;
\echo

-- creating extensions will fire superuser evtrigs
set role privileged_role;
create extension postgres_fdw;
drop extension postgres_fdw;
\echo

-- creating fdws will fire superuser evtrigs
create foreign data wrapper new_fdw;
drop foreign data wrapper new_fdw;
\echo

-- creating publications will fire superuser evtrigs
create publication p for all tables;
drop publication p;
\echo

-- a non-privileged role cannot alter a superuser owned evtrig
set role rolecreator;
\echo

alter event trigger event_trigger_1 disable;
\echo

-- the privileged role cannot alter a superuser owned evtrig
set role privileged_role;
\echo

alter event trigger event_trigger_1 disable;
\echo

-- a reserved role cannot alter a superuser owned evtrig
set role supabase_storage_admin;
\echo

alter event trigger event_trigger_1 disable;
\echo

-- only the superuser can alter its own event triggers
set role postgres;
alter event trigger event_trigger_1 disable;
\echo

-- a non-privileged role cannot drop a superuser owned evtrig
set role rolecreator;
\echo

drop event trigger event_trigger_1;
\echo

-- the privileged role cannot drop a superuser owned evtrig
set role privileged_role;
\echo

drop event trigger event_trigger_1;
\echo

-- a reserved role cannot drop a superuser owned evtrig
set role supabase_storage_admin;
\echo

drop event trigger event_trigger_1;
\echo

-- only the superuser can drop its own event triggers
set role postgres;
drop event trigger event_trigger_1;
\echo

-- disable logging event triggers
reset supautils.log_skipped_evtrigs;
