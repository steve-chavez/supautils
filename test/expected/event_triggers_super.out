-- skipped event triggers are logged
set supautils.log_skipped_evtrigs = true;
\echo

-- a superuser can create an event trigger
set role postgres;
\echo

create function super_show_current_user()
    returns event_trigger
    language plpgsql as
$$
begin
    raise notice 'Executing superuser event trigger: current_user is %', current_user;
end;
$$;
\echo

create event trigger event_trigger_1 on ddl_command_end
execute procedure super_show_current_user();
\echo

-- the event trigger is owned by the superuser
select evtowner::regrole from pg_event_trigger where evtname = 'event_trigger_1';
 evtowner 
----------
 postgres
(1 row)

-- the superuser evtrig will execute for superuser
create table super_thing();
NOTICE:  Executing superuser event trigger: current_user is postgres
\echo

-- another superuser can create an event trigger
create role super2 superuser;
set role super2;
create function super2_show_current_user()
    returns event_trigger
    language plpgsql as
$$
begin
    raise notice 'Executing super2 event trigger: current_user is %', current_user;
end;
$$;
NOTICE:  Skipping event trigger function "super_show_current_user" for user "super2"
DETAIL:  "super2" is a superuser and the function "super_show_current_user" is not owned by the same role, it's owned by "postgres"
create event trigger super2_event_trigger on ddl_command_end
execute procedure super2_show_current_user();
\echo

-- only one superuser evtrig will execute for superuser (the one he owns)
create table super2_thing();
NOTICE:  Skipping event trigger function "super_show_current_user" for user "super2"
DETAIL:  "super2" is a superuser and the function "super_show_current_user" is not owned by the same role, it's owned by "postgres"
NOTICE:  Executing super2 event trigger: current_user is super2
\echo

-- super2 evtrigs won't run here, only the postgres owned evtrigs
set role postgres;
create table super_duper_thing();
NOTICE:  Executing superuser event trigger: current_user is postgres
NOTICE:  Skipping event trigger function "super2_show_current_user" for user "postgres"
DETAIL:  "postgres" is a superuser and the function "super2_show_current_user" is not owned by the same role, it's owned by "super2"
\echo

-- super2 cleanup
drop event trigger super2_event_trigger;
drop function super2_show_current_user;
NOTICE:  Executing superuser event trigger: current_user is postgres
drop table super2_thing;
NOTICE:  Executing superuser event trigger: current_user is postgres
drop role super2;
\echo

-- A regular user is not able to create event trigger using a superuser function
set role privileged_role;
\echo

create event trigger event_trigger_2 on ddl_command_end
execute procedure super_show_current_user();
ERROR:  Non-superuser owned event trigger must execute a non-superuser owned function
DETAIL:  The current user "privileged_role" is not a superuser and the function "super_show_current_user" is owned by a superuser
\echo

-- A reserved role will execute the superuser evtrig
set role supabase_storage_admin;
\echo

create table storage_stuff();
NOTICE:  Executing superuser event trigger: current_user is supabase_storage_admin
\echo

drop table storage_stuff;
NOTICE:  Executing superuser event trigger: current_user is supabase_storage_admin
\echo

-- creating extensions will fire superuser evtrigs
set role privileged_role;
create extension postgres_fdw;
NOTICE:  Executing superuser event trigger: current_user is postgres
drop extension postgres_fdw;
NOTICE:  Executing superuser event trigger: current_user is postgres
\echo

-- creating fdws will fire superuser evtrigs
create foreign data wrapper new_fdw;
NOTICE:  Executing superuser event trigger: current_user is postgres
drop foreign data wrapper new_fdw;
NOTICE:  Executing superuser event trigger: current_user is privileged_role
\echo

-- creating publications will fire superuser evtrigs
create publication p for all tables;
NOTICE:  Executing superuser event trigger: current_user is postgres
drop publication p;
NOTICE:  Executing superuser event trigger: current_user is privileged_role
\echo

-- a non-privileged role cannot alter a superuser owned evtrig
set role rolecreator;
\echo

alter event trigger event_trigger_1 disable;
ERROR:  must be owner of event trigger event_trigger_1
\echo

-- the privileged role cannot alter a superuser owned evtrig
set role privileged_role;
\echo

alter event trigger event_trigger_1 disable;
ERROR:  must be owner of event trigger event_trigger_1
\echo

-- a reserved role cannot alter a superuser owned evtrig
set role supabase_storage_admin;
\echo

alter event trigger event_trigger_1 disable;
ERROR:  must be owner of event trigger event_trigger_1
\echo

-- only the superuser can alter its own event triggers
set role postgres;
alter event trigger event_trigger_1 disable;
\echo

-- a non-privileged role cannot drop a superuser owned evtrig
set role rolecreator;
\echo

drop event trigger event_trigger_1;
ERROR:  must be owner of event trigger event_trigger_1
\echo

-- the privileged role cannot drop a superuser owned evtrig
set role privileged_role;
\echo

drop event trigger event_trigger_1;
ERROR:  must be owner of event trigger event_trigger_1
\echo

-- a reserved role cannot drop a superuser owned evtrig
set role supabase_storage_admin;
\echo

drop event trigger event_trigger_1;
ERROR:  must be owner of event trigger event_trigger_1
\echo

-- only the superuser can drop its own event triggers
set role postgres;
drop event trigger event_trigger_1;
\echo

-- disable logging event triggers
reset supautils.log_skipped_evtrigs;
